### PR TITLE
Kernel+LibDebug+HackStudio: Implement variable watchpoints

### PR DESCRIPTION
**Pull request by FalseHonesty**
This PR adds the ability to place a hardware watchpoint on a debugged variable. To do this from HackStudio, just right click on a variable in the variables view and select add watchpoint. Under the hood this will use ptrace to communicate with the kernel to modify the proper x86 debug registers.

This is my first time ever looking at the Kernel in depth, much less modifying it, so I'm definitely open to suggestions there.

### CODE REVIEW

**Comment by FalseHonesty** [comment]
cc @itamar8910

File: Kernel/Arch/i386/CPU.cpp [hunk 1]
```diff
@@ -1390,6 +1530,9 @@ extern "C" void enter_thread_context(Thread* from_thread, Thread* to_thread)
     set_fs(to_tss.fs);
     set_gs(to_tss.gs);
 
+    read_debug_registers_into(from_thread->debug_register_state());
```
Comment (bgianfo) [inline Kernel/Arch/i386/CPU.cpp:1533]: Should we guard this case so it's only performed if the processes involved are being traced?

File: Kernel/Arch/i386/CPU.cpp [hunk 2]
```diff
@@ -854,6 +855,114 @@ FlatPtr read_cr4()
...
+#else
+    asm volatile("mov %%rax, %%dr2" ::"a"(value));
+#endif
+}
+
+FlatPtr read_dr3()
+{
+    FlatPtr dr3;
+#if ARCH(I386)
+    asm("mov %%dr3, %%eax"
```
Comment (bgianfo) [inline Kernel/Arch/i386/CPU.cpp:948]: Given these are all the same besides the debug register name, perhaps we should generate their definition via macros?

File: Kernel/Syscalls/ptrace.cpp [hunk 3]
```diff
@@ -229,4 +242,56 @@ KResult Process::poke_user_data(Userspace<u32*> address, u32 data)
...
+        break;
+    case 3:
+        m_debug_register_state.dr3 = data;
+        break;
+    case 7:
+        m_debug_register_state.dr7 = data;
+        break;
+    default:
+        return EFAULT;
+    }
```
Comment (bgianfo) [inline Kernel/Syscalls/ptrace.cpp:293]: Should we return `EINVAL`? As we don't really fault here?

File: Kernel/Arch/i386/CPU.cpp [hunk 4]
```diff
@@ -854,6 +855,114 @@ FlatPtr read_cr4()
...
+{
+    debug_register_state.dr0 = read_dr0();
+    debug_register_state.dr1 = read_dr1();
+    debug_register_state.dr2 = read_dr2();
+    debug_register_state.dr3 = read_dr3();
+    debug_register_state.dr6 = read_dr6();
+    debug_register_state.dr7 = read_dr7();
+}
+
+void write_debug_registers_from(DebugRegisterState& state)
```
Comment (bgianfo) [inline Kernel/Arch/i386/CPU.cpp:868]: This should be a const reference?

File: Userland/Libraries/LibDebug/DebugSession.h [hunk 5]
```diff
@@ -223,6 +238,45 @@ void DebugSession::run(DesiredInitialDebugeeState initial_debugee_state, Callbac
             state = State::FreeRun;
 
         auto regs = get_registers();
+
+        auto debug_status = peek_debug(6);
```
Comment (bgianfo) [inline Userland/Libraries/LibDebug/DebugSession.h:242]: Is there a constant we can use instead of `6`?

Comment (FalseHonesty) [reply Userland/Libraries/LibDebug/DebugSession.h]: I've moved them to constants in the ptrace header, that feels like the most applicable place to put them?

File: Kernel/Syscalls/ptrace.cpp [hunk 6]
```diff
@@ -229,4 +242,56 @@ KResult Process::poke_user_data(Userspace<u32*> address, u32 data)
...
+        data = m_debug_register_state.dr3;
+        break;
+    case 6:
+        data = m_debug_register_state.dr6;
+        break;
+    case 7:
+        data = m_debug_register_state.dr7;
+        break;
+    default:
+        return EFAULT;
```
Comment (bgianfo) [inline Kernel/Syscalls/ptrace.cpp:268]: EINVAL here as well?

**Review by bgianfo — APPROVED** [review]

**Comment by awesomekling** [comment]
Hmm wait, how does the kernel preserve per-thread debug registers when a context switch occurs?

**Comment by FalseHonesty** [comment]
> Hmm wait, how does the kernel preserve per-thread debug registers when a context switch occurs?

Relevant code is [here](https://github.com/SerenityOS/serenity/pull/6353/files#diff-05c65ddc0798c85b83113c2bd6c1285bcd8f7cf42843a7c79faa8cf86bc65acfR1445). When I first implemented debug registers, I didn't realize I had to manually store/load on context switch, so everything was crashing. After looking for a while where to put it, I found this method (`enter_thread_context`) where other states were loaded/saved and realized it had a comment saying to add support for debug registers, so I figure it's the right location?

**Pull request merged**